tests/rados/test_bluestore_features.py

```diff
@@ -0,0 +1,80 @@
+"""
+This file contains the  methods to verify the Bluestore features.
+"""
+
+
+import time
+
+from ceph.ceph_admin import CephAdmin
+from ceph.rados.core_workflows import RadosOrchestrator
+from ceph.rados.rados_scrub import RadosScrubber
+from tests.rados.monitor_configurations import MonConfigMethods
+from utility.log import Log
+from utility.utils import method_should_succeed
+
+log = Log(__name__)
+
+
+def run(ceph_cluster, **kw):
+    """
+    Performs various pool related validation tests
+    Returns:
+        1 -> Fail, 0 -> Pass
+    """
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_object = RadosOrchestrator(node=cephadm)
+    mon_obj = MonConfigMethods(rados_obj=rados_object)
+    scrub_obj = RadosScrubber(node=cephadm)
+    target_configs = config["cache_trim_max_skip_pinned"]["configurations"]
+
+    if config.get("cache_trim_max_skip_pinned"):
+        if not mon_obj.set_config(
+            section="osd", name="bluestore_cache_trim_max_skip_pinned", value="1000"
+        ):
+            log.error("Error setting config ")
+            return 1
+
+        # Creating pools and starting the test
+        for entry in target_configs.values():
+            log.debug(
+                f"Creating {entry['pool_type']} pool on the cluster with name {entry['pool_name']}"
+            )
+        method_should_succeed(
+            rados_object.create_pool,
+            **entry,
+        )
+        rados_object.bench_write(**entry)
+
+        # Get the scrub and deep scrup time stamp before scrub and deep scrub execution
+        before_scrub_log = scrub_obj.get_pg_dump("pgid", "last_scrub_stamp")
+        before_deep_scrub_log = scrub_obj.get_pg_dump("pgid", "last_deep_scrub_stamp")
+
+        # performing scrub and deep-scrub
+        rados_object.run_scrub()
+        rados_object.run_deep_scrub()
+        scrub_status = 1
+        deep_scrub_status = 1
+
+        while scrub_status == 1 and deep_scrub_status == 1:
+            time.sleep(10)
+            after_scrub_log = scrub_obj.get_pg_dump("pgid", "last_scrub_stamp")
+            after_deep_scrub_log = scrub_obj.get_pg_dump(
+                "pgid", "last_deep_scrub_stamp"
+            )
+            scrub_status = scrub_obj.verify_scrub_deepscrub(
+                before_scrub_log, after_scrub_log, "scrub"
+            )
+            deep_scrub_status = scrub_obj.verify_scrub_deepscrub(
+                before_deep_scrub_log, after_deep_scrub_log, "deepscrub"
+            )
+            log.info("scrubbing and deep-scrubbing are in progress")
+        log.info("scrubbing and deep-scrubbing completed")
+
+        # check fo r crashes in the cluster
```

Comment:
We need to continuously monitor the OSD memory growth behaviour at this point, when scrubbing is in progress.

The root cause for the test here is the memory growth, when it crosses a certain threshold, the OSD daemons would crash.

we need to adda method to check the memory usage of OSDs with scrubbing and IO's running. With ` bluestore_cache_trim_max_skip_pinned ` set to 1000 by default, we should see very less growth in OSD memory consumption.

Reply:
Ref : https://bugzilla.redhat.com/show_bug.cgi?id=1931504#c5

Reply:
I checking  the default  bluestore_cache_trim_max_skip_pinned value with 1000.Test case will fail if it is not equal to 1000.

Reply:
Please add the check for OSD memory growth as well. getting the default value for 'bluestore_cache_trim_max_skip_pinned' and verifying seems very low ROI and not enough validation.

Reply:
Checking the memory growth while scrubbing.For every 10 seconds checking the total_used_bytes of the cluster. If the memory growth is more than 10% I am failing the test case.The code exist from line numbers 67 to 82.

Reply:
Hey Bharath, I think there is a confusion here. We are supposed to check the memory consumption grwoth of the OSD daemons, not the ceph cluster storage stats. 

Can you please check again? Similar activity is needed for task : RHCEPHQE-7917 . Ref : https://access.redhat.com/articles/351143

Reply:
Calculating the memory used by every OSD by using the heap dump command. The sample output of heap dump output is as below-
_-----------------------------------------------
MALLOC:      538211456 (  513.3 MiB) Bytes in use by application
MALLOC: +            0 (    0.0 MiB) Bytes in page heap freelist
MALLOC: +     14341008 (   13.7 MiB) Bytes in central cache freelist
MALLOC: +      4471808 (    4.3 MiB) Bytes in transfer cache freelist
MALLOC: +     19725296 (   18.8 MiB) Bytes in thread cache freelists
MALLOC: +      3670016 (    3.5 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =    580419584 (  553.5 MiB) Actual memory used (physical + swap)
MALLOC: +    197885952 (  188.7 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =    778305536 (  742.2 MiB) Virtual address space used
MALLOC:
MALLOC:          18491              Spans in use
MALLOC:             34              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------_ 

Executing the tests for 45 minutes and every 10 seconds the details are storing in a list.Once the data is retrieved the relative growth is calculating for every OSD.If the growth is greater than 80% then making testcase fail. 

To get the 80% benchmark , I executed the tests with the 256KB and 64MB.For 256KB the memory growth is very minimal so I choose the 64MB.In addition to that, executed the tests with the low "bluestore_cache_trim_max_skip_pinned" value that is 10 and noticed that the memory growth some time more than 100% but with the default value that growth is in between 60-80%.

**Memory growth when bluestore_cache_trim_max_skip_pinned=10 & objectsize=64MB:**
2023-04-08 19:21:16,853 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.0 consuming more memory with the memory growth 44.58713547406898
2023-04-08 19:21:16,854 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.0 is 44.58713547406898 
2023-04-08 19:21:16,856 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.3 consuming more memory with the memory growth 145.84464591690133
2023-04-08 19:21:16,857 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.3 is 145.84464591690133 
2023-04-08 19:21:16,858 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.6 consuming more memory with the memory growth 41.93759008922337
2023-04-08 19:21:16,859 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.6 is 41.93759008922337 
2023-04-08 19:21:16,860 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.9 consuming more memory with the memory growth 53.49275179438129
2023-04-08 19:21:16,861 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.9 is 53.49275179438129 
2023-04-08 19:21:16,862 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.12 consuming more memory with the memory growth 45.60211673729597
2023-04-08 19:21:16,863 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.12 is 45.60211673729597 
2023-04-08 19:21:16,864 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.2 consuming more memory with the memory growth 38.50719602435498
2023-04-08 19:21:16,864 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.2 is 38.50719602435498 
2023-04-08 19:21:16,865 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.5 consuming more memory with the memory growth 68.27855738178029
2023-04-08 19:21:16,866 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.5 is 68.27855738178029 
2023-04-08 19:21:16,866 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.8 consuming more memory with the memory growth 32.06705562117973
2023-04-08 19:21:16,867 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.8 is 32.06705562117973 
2023-04-08 19:21:16,868 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.11 consuming more memory with the memory growth 45.774732483609924
2023-04-08 19:21:16,868 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.11 is 45.774732483609924 
2023-04-08 19:21:16,869 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.14 consuming more memory with the memory growth 30.69973365755619
2023-04-08 19:21:16,869 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.14 is 30.69973365755619 
2023-04-08 19:21:16,870 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.1 consuming more memory with the memory growth 26.564678005489018
2023-04-08 19:21:16,871 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.1 is 26.564678005489018 
2023-04-08 19:21:16,871 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.4 consuming more memory with the memory growth 68.50086699397161
2023-04-08 19:21:16,872 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.4 is 68.50086699397161 
2023-04-08 19:21:16,873 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.7 consuming more memory with the memory growth 55.005894388812855
2023-04-08 19:21:16,873 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.7 is 55.005894388812855 
2023-04-08 19:21:16,874 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.10 consuming more memory with the memory growth 49.51864084399263
2023-04-08 19:21:16,875 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.10 is 49.51864084399263 
2023-04-08 19:21:16,875 (cephci.test_bluestore_features) [ERROR] - cephci.tests.rados.test_bluestore_features.py:110 - The osd.13 consuming more memory with the memory growth 54.60954384746342
2023-04-08 19:21:16,876 (cephci.test_bluestore_features) [INFO] - cephci.tests.rados.test_bluestore_features.py:112 - The memory growth for the osd.13 is 54.60954384746342

Log files:
   1.http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-WQTRYO
   2.http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-JVYXEL